OpenRA.Game/Network/Connection.cs

```diff
@@ -131,57 +136,66 @@ public void Dispose()
 
 	sealed class NetworkConnection : EchoConnection
 	{
-		TcpClient socket;
-		int clientId;
-		ConnectionState connectionState = ConnectionState.Connecting;
-		Thread t;
+		readonly TcpClient tcp;
+		readonly List<byte[]> queuedSyncPackets = new List<byte[]>();
+		volatile ConnectionState connectionState = ConnectionState.Connecting;
+		volatile int clientId = -1;
+		bool disposed;
 
 		public NetworkConnection(string host, int port)
 		{
-			t = new Thread(_ =>
+			try
 			{
-				try
+				tcp = new TcpClient(host, port) { NoDelay = true };
+				new Thread(NetworkConnectionReceive)
 				{
-					socket = new TcpClient(host, port);
-					socket.NoDelay = true;
-					var reader = new BinaryReader(socket.GetStream());
-					var serverProtocol = reader.ReadInt32();
-
-					if (ProtocolVersion.Version != serverProtocol)
-						throw new InvalidOperationException(
-							"Protocol version mismatch. Server={0} Client={1}"
-								.F(serverProtocol, ProtocolVersion.Version));
-
-					clientId = reader.ReadInt32();
-					connectionState = ConnectionState.Connected;
-
-					for (;;)
-					{
-						var len = reader.ReadInt32();
-						var client = reader.ReadInt32();
-						var buf = reader.ReadBytes(len);
-						if (len == 0)
-							throw new NotImplementedException();
-						lock (this)
-							receivedPackets.Add(new ReceivedPacket { FromClient = client, Data = buf });
-					}
-				}
-				catch { }
-				finally
+					Name = GetType().Name + " " + host + ":" + port,
+					IsBackground = true
+				}.Start(tcp.GetStream());
+			}
+			catch { }
+			finally
+			{
+				connectionState = ConnectionState.NotConnected;
```

Comment:
It looks like the code hasn't changed since these comments, is this still an issue?